File: app/src/main/java/closer/vlllage/com/closer/handler/SearchGroupHandler.java
package closer.vlllage.com.closer.handler;

import java.util.ArrayList;
import java.util.List;

import closer.vlllage.com.closer.handler.helpers.Val;
import closer.vlllage.com.closer.handler.group.SearchGroupsAdapter;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.store.StoreHandler;
import closer.vlllage.com.closer.store.models.Group;
import closer.vlllage.com.closer.store.models.GroupAction;
import closer.vlllage.com.closer.store.models.GroupAction_;

public class SearchGroupHandler extends PoolMember {

    private String searchQuery = "";
    private SearchGroupsAdapter searchGroupsAdapter;
    private List<Group> groupsCache;
    private boolean hideCreateGroupOption;

    public void showGroupsForQuery(SearchGroupsAdapter searchGroupsAdapter, String searchQuery) {
        this.searchQuery = searchQuery.toLowerCase();
        this.searchGroupsAdapter = searchGroupsAdapter;

        if (!hideCreateGroupOption) {
            searchGroupsAdapter.setCreatePublicGroupName(searchQuery.trim().isEmpty() ? null : searchQuery);
        }

        if (this.groupsCache != null) {
            setGroups(this.groupsCache);
        }
    }

    public void setGroups(List<Group> allGroups) {
        this.groupsCache = allGroups;

        List<Group> groups = new ArrayList<>();
        for(Group group : allGroups) {
            if (group.getName() != null) {
                if (group.getName().toLowerCase().contains(searchQuery) ||
                        $(Val.class).of(group.getAbout(), "").toLowerCase().contains(searchQuery) ||
                        groupActionNamesContains(group, searchQuery)) {
                    groups.add(group);
                }
            }
        }

        searchGroupsAdapter.setGroups(groups);
    }

    private boolean groupActionNamesContains(Group group, String searchQuery) {
        List<GroupAction> groupActions = $(StoreHandler.class).getStore().box(GroupAction.class).query()
                .equal(GroupAction_.group, group.getId()).build().find();

        for (GroupAction groupAction : groupActions) {
            if (groupAction.getName().toLowerCase().contains(searchQuery)) {
                return true;
            }
        }

        return false;
    }

    public void hideCreateGroupOption() {
        this.hideCreateGroupOption = true;
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/group/SearchGroupHandler.kt
package closer.vlllage.com.closer.handler.group

import closer.vlllage.com.closer.handler.helpers.Val
import closer.vlllage.com.closer.store.StoreHandler
import closer.vlllage.com.closer.store.models.Group
import closer.vlllage.com.closer.store.models.GroupAction
import closer.vlllage.com.closer.store.models.GroupAction_
import com.queatz.on.On
import io.objectbox.query.QueryBuilder
import io.reactivex.rxjava3.subjects.BehaviorSubject
import java.util.*

class SearchGroupHandler constructor(private val on: On) {

    private var searchQuery = ""
    private var groupsCache: List<Group>? = null

    val groups = BehaviorSubject.create<List<Group>>()
    val createGroupName = BehaviorSubject.createDefault("")

    fun showGroupsForQuery(searchQuery: String) {
        if (this.searchQuery == searchQuery.toLowerCase(Locale.getDefault())) return

        this.searchQuery = searchQuery.toLowerCase(Locale.getDefault())

        createGroupName.onNext(if (searchQuery.isBlank()) "" else searchQuery)

        groupsCache?.let { setGroups(it) }
    }

    fun setGroups(allGroups: List<Group>, includeTopics: Boolean = false) {
        this.groupsCache = allGroups

        if (includeTopics) {
            this.groups.onNext(allGroups)

            return
        }

        val groups = mutableListOf<Group>()
        for (group in allGroups) {
            if (group.name != null) {
                if (group.name!!.toLowerCase(Locale.getDefault()).contains(searchQuery) ||
                        on<Val>().of(group.about, "").toLowerCase(Locale.getDefault()).contains(searchQuery) ||
                        groupActionNamesContains(group, searchQuery)) {
                    groups.add(group)
                }
            }
        }

        this.groups.onNext(groups)
    }

    private fun groupActionNamesContains(group: Group, searchQuery: String): Boolean {
        val groupActions = on<StoreHandler>().store.box(GroupAction::class).query()
            .equal(GroupAction_.group, group.id!!, QueryBuilder.StringOrder.CASE_SENSITIVE)
            .build()
            .find()

        for (groupAction in groupActions) {
            if (groupAction.name!!.toLowerCase(Locale.getDefault()).contains(searchQuery)) {
                return true
            }
        }

        return false
    }
}
